<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <title>Title</title>
</head>
<body>
    <script>
        Function.prototype.myCall = function (context) {
            var context = context || window;
            context.fn = this;

            var args = [];
            for(var i = 1, len = arguments.length; i < len; i++) {
                args.push('arguments[' + i + ']');
                //args.push(arguments[i]); //存在bug，
            }

            var result = eval('context.fn(' + args +')'); //注意，这里隐式调用了args数组的toString方法，将数组各成员以 逗号, 拼接

            delete context.fn
            return result;
        }




        function fn1(){
            console.log(this);
            console.log(1);
        }
        function fn2(){
            console.log(this);
            console.log(2);
        }
        function fn3(n1,n2){
            console.log(this);
            console.log(n1+n2);
        }
        fn1.myCall(fn2);// fn2 , 1
        fn1.myCall.myCall(fn2); //window ,2
        fn1.myCall.myCall.myCall(fn2); //window ,2
        fn3.myCall(fn2,3,4);// fn2 , 7
		
		
		
		
	var obj = {
        value: 1
    }

    function bar(name, age) {
        console.log(this.value);
        return {
            value: this.value,
            name: name,
            age: age
        }
    }

	// myCall 中采用 args.push(arguments[i])方式存储参数，调用 eval时
	// 变成 eval(" obj.fn(kevin,18)"), 其中 obj.fn 指向函数 bar
	// eval执行时，会查找kevin 变量，出现错误。
    console.log(bar.call2(obj, 'kevin', 18)); 
    </script>
</body>
</html>